# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/libwebrtcglue/VideoConduit.cpp
# Commit: 7057437ddf19
# Full Hash: 7057437ddf194b62a158ce84345802e8b2bc6ca8
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 6
# Description:
#   Bug 1654112 - Constify VideoConduit members that are based on prefs. r=ng
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D112848
# ==============================================================================

diff -r 01569f2b1141 -r 7057437ddf19 dom/media/webrtc/libwebrtcglue/VideoConduit.cpp
--- a/dom/media/webrtc/libwebrtcglue/VideoConduit.cpp	Tue Apr 06 18:29:30 2021 +0200
+++ b/dom/media/webrtc/libwebrtcglue/VideoConduit.cpp	Wed Apr 07 09:44:33 2021 +0200
@@ -289,7 +289,7 @@
  */
 RefPtr<VideoSessionConduit> VideoSessionConduit::Create(
     RefPtr<WebrtcCallWrapper> aCall, nsCOMPtr<nsISerialEventTarget> aStsThread,
-    std::string aPCHandle) {
+    Options aOptions, std::string aPCHandle) {
   MOZ_ASSERT(NS_IsMainThread());
   MOZ_ASSERT(aCall, "missing required parameter: aCall");
   CSFLogVerbose(LOGTAG, "%s", __FUNCTION__);
@@ -298,8 +298,8 @@
     return nullptr;
   }
 
-  auto obj =
-      MakeRefPtr<WebrtcVideoConduit>(aCall, aStsThread, std::move(aPCHandle));
+  auto obj = MakeRefPtr<WebrtcVideoConduit>(
+      aCall, aStsThread, std::move(aOptions), std::move(aPCHandle));
   if (obj->Init() != kMediaConduitNoError) {
     CSFLogError(LOGTAG, "%s VideoConduit Init Failed ", __FUNCTION__);
     return nullptr;
@@ -310,7 +310,7 @@
 
 WebrtcVideoConduit::WebrtcVideoConduit(
     RefPtr<WebrtcCallWrapper> aCall, nsCOMPtr<nsISerialEventTarget> aStsThread,
-    std::string aPCHandle)
+    Options aOptions, std::string aPCHandle)
     : mTransportMonitor("WebrtcVideoConduit"),
       mStsThread(aStsThread),
       mMutex("WebrtcVideoConduit::mMutex"),
@@ -322,6 +322,15 @@
       mEngineTransmitting(false),
       mEngineReceiving(false),
       mSendingFramerate(DEFAULT_VIDEO_MAX_FRAMERATE),
+      mVideoLatencyTestEnable(aOptions.mVideoLatencyTestEnable),
+      mMinBitrate(aOptions.mMinBitrate),
+      mStartBitrate(aOptions.mStartBitrate),
+      mPrefMaxBitrate(aOptions.mPrefMaxBitrate),
+      mMinBitrateEstimate(aOptions.mMinBitrateEstimate),
+      mDenoising(aOptions.mDenoising),
+      mLockScaling(aOptions.mLockScaling),
+      mSpatialLayers(aOptions.mSpatialLayers),
+      mTemporalLayers(aOptions.mTemporalLayers),
       mActiveCodecMode(webrtc::VideoCodecMode::kRealtimeVideo),
       mCodecMode(webrtc::VideoCodecMode::kRealtimeVideo),
       mCall(aCall),
@@ -869,93 +878,17 @@
   return mCall->Call()->GetStats();
 }
 
-MediaConduitErrorCode WebrtcVideoConduit::InitMain() {
+MediaConduitErrorCode WebrtcVideoConduit::Init() {
   MOZ_ASSERT(NS_IsMainThread());
 
-  nsresult rv;
-  nsCOMPtr<nsIPrefService> prefs =
-      do_GetService("@mozilla.org/preferences-service;1", &rv);
-  if (!NS_WARN_IF(NS_FAILED(rv))) {
-    nsCOMPtr<nsIPrefBranch> branch = do_QueryInterface(prefs);
+  CSFLogDebug(LOGTAG, "%s this=%p", __FUNCTION__, this);
 
-    if (branch) {
-      int32_t temp;
-      Unused << NS_WARN_IF(NS_FAILED(branch->GetBoolPref(
-          "media.video.test_latency", &mVideoLatencyTestEnable)));
-      if (!NS_WARN_IF(NS_FAILED(branch->GetIntPref(
-              "media.peerconnection.video.min_bitrate", &temp)))) {
-        if (temp >= 0) {
-          mMinBitrate = KBPS(temp);
-        }
-      }
-      if (!NS_WARN_IF(NS_FAILED(branch->GetIntPref(
-              "media.peerconnection.video.start_bitrate", &temp)))) {
-        if (temp >= 0) {
-          mStartBitrate = KBPS(temp);
-        }
-      }
-      if (!NS_WARN_IF(NS_FAILED(branch->GetIntPref(
-              "media.peerconnection.video.max_bitrate", &temp)))) {
-        if (temp >= 0) {
-          mPrefMaxBitrate = KBPS(temp);
-        }
-      }
-      if (mMinBitrate != 0 && mMinBitrate < kViEMinCodecBitrate_bps) {
-        mMinBitrate = kViEMinCodecBitrate_bps;
-      }
-      if (mStartBitrate < mMinBitrate) {
-        mStartBitrate = mMinBitrate;
-      }
-      if (mPrefMaxBitrate && mStartBitrate > mPrefMaxBitrate) {
-        mStartBitrate = mPrefMaxBitrate;
-      }
-      // XXX We'd love if this was a live param for testing adaptation/etc
-      // in automation
-      if (!NS_WARN_IF(NS_FAILED(branch->GetIntPref(
-              "media.peerconnection.video.min_bitrate_estimate", &temp)))) {
-        if (temp >= 0) {
-          mMinBitrateEstimate = temp;  // bps!
-        }
-      }
-      if (!NS_WARN_IF(NS_FAILED(branch->GetIntPref(
-              "media.peerconnection.video.svc.spatial", &temp)))) {
-        if (temp >= 0) {
-          mSpatialLayers = temp;
-        }
-      }
-      if (!NS_WARN_IF(NS_FAILED(branch->GetIntPref(
-              "media.peerconnection.video.svc.temporal", &temp)))) {
-        if (temp >= 0) {
-          mTemporalLayers = temp;
-        }
-      }
-      Unused << NS_WARN_IF(NS_FAILED(branch->GetBoolPref(
-          "media.peerconnection.video.denoising", &mDenoising)));
-      Unused << NS_WARN_IF(NS_FAILED(branch->GetBoolPref(
-          "media.peerconnection.video.lock_scaling", &mLockScaling)));
-    }
-  }
 #ifdef MOZ_WIDGET_ANDROID
   if (mozilla::camera::VideoEngine::SetAndroidObjects() != 0) {
     CSFLogError(LOGTAG, "%s: could not set Android objects", __FUNCTION__);
     return kMediaConduitSessionNotInited;
   }
 #endif  // MOZ_WIDGET_ANDROID
-  return kMediaConduitNoError;
-}
-
-/**
- * Performs initialization of the MANDATORY components of the Video Engine
- */
-MediaConduitErrorCode WebrtcVideoConduit::Init() {
-  MOZ_ASSERT(NS_IsMainThread());
-
-  CSFLogDebug(LOGTAG, "%s this=%p", __FUNCTION__, this);
-  MediaConduitErrorCode result;
-  result = InitMain();
-  if (result != kMediaConduitNoError) {
-    return result;
-  }
 
   CSFLogDebug(LOGTAG, "%s Initialization Done", __FUNCTION__);
   return kMediaConduitNoError;